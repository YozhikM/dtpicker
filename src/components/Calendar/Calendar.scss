body {
  font-family: 'Rubik';
}
.root {
  display: inline-block;
  width: 273px;
  z-index: 1;
  position: relative;
  bottom: 28px;

   table {
     border-collapse: collapse;
   }

   tr {
     display: flex;
   }

   tr:nth-of-type(1) {
     justify-content: flex-end;
   }

  td, th {
    width: 39px;
    height: 39px;
    text-align: center;
    line-height: 39px;
    padding: 0;
    box-sizing: border-box;
    cursor: pointer;
    outline: none;
   }
}

.button_container {
  display: flex;
  width: 600px;
  justify-content: space-around;

  button {
    width: 273px;
    display: block;
    text-align: center;
    margin: 3% 0 3% 0;
    color: #3c3f40;
    font-size: 18px;
    font-weight: bold;
    cursor: pointer;
    border: none;
    background-color: transparent;
    outline: none;
  }
}

.button_single_container {
  position: absolute;
  width: 273px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 45px;
  z-index: -1;

  button {
    display: block;
    text-align: center;
    margin: 0;
    color: #3c3f40;
    font-size: 18px;
    font-weight: bold;
    cursor: pointer;
    border: none;
    background-color: transparent;
    outline: none;
    transform: scale(1);
    transition: transform .6s ease;
  }
  button:hover {
    transform: scale(.8);
    transition: transform .6s ease;
  }
}
  .left_arrow {
    position: relative;
    display: inline-block !important;
    right: 120px;
    top: 36px;
    z-index: 2;
    color: #333 !important;
    margin: 0 !important;
    border: none !important;
    background-color: transparent !important;
    padding: 0 !important;
  }
  .right_arrow {
    position: relative;
    left: 117px;
    display: inline-block !important;
    top: 8.5px;
    z-index: 2;
    color: #333 !important;
    margin: 0 !important;
    border: none !important;
    background-color: transparent !important;
    padding: 0 !important;
}
.calendar_container {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 273px;

  span {
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
    margin: 3% auto;
    color: #3c3f40;
    font-size: 18px;
    font-weight: bold;
    cursor: pointer;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
  }

  button {
    text-align: center;
    font-size: 18px;
    color: #333;
    cursor: pointer;
    background-color: #e4e7e7;
    border: 1px solid #e4e7e7;
    border-radius: 5px;
    display: block;
    margin: 3% auto;
    padding: 1.5%;
  }
}
.table {
  width: 273px;
  display: inline-block;
  z-index: 2;

  span {
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
    margin: 3% auto;
    color: #3c3f40;
    font-size: 18px;
    font-weight: bold;
    cursor: pointer;
    background-color: #fff;
  }
}
.c_button_container {
  display: flex;
  width: 273px;
  justify-content: space-between;

  button {
    display: block;
    text-align: center;
    margin: 3% 0 3% 0;
    color: #3c3f40;
    font-size: 18px;
    font-weight: bold;
    cursor: pointer;
    border: none;
    background-color: transparent;
    outline: none;
    z-index: 1;
  }
}

  .time_picker {
    width: 273px;
    text-align: center;

    button {
      cursor: pointer;
      border: none;
      background-color: transparent;
      outline: none;
      margin: 0 auto;
      padding: 0 !important;
    }
  }
