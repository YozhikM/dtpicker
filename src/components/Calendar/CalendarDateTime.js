/* @flow */

import React from 'react';
import CalendarMonth from './CalendarMonth';
import CalendarTimePicker from './CalendarTimePicker';
import s from './Calendar.scss';
import { format } from 'date-fns';

type Props = {
  date: Date,
  activeDates: Date,
  time?: boolean,
  leftArrow?: boolean,
  rightArrow?: boolean,
  onClickDay?: (date: Date) => void,
  onChangeDate?: (date: Date) => void
};

type State = {
  date: Date
};

class CalendarSingle extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      date: this.props.date
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.date !== this.props.date) {
      this.setState({ date: nextProps.date });
    }
  }

  onChangeDate = () => {
    const { onChangeDate } = this.props;
    if (onChangeDate) onChangeDate(this.state.date);
  };

  render() {
    const { date } = this.state;
    const { time, activeDates } = this.props;
    return (
      <div className={s.calendar_container}>
        <p>{format((this.props.date), 'hh:mm dd/MM/YY')}</p>
        <CalendarMonth {...this.props} date={date} leftArrow rightArrow />

        {time && (
          <CalendarTimePicker
            activeDates={activeDates}
            date={date}
            onChangeDate={this.onChangeDate}
          />
        )}
      </div>
    );
  }
}

export default CalendarSingle;
