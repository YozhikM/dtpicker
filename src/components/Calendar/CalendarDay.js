/* @flow */

import React from 'react';
import { getDate } from 'date-fns';

type Props = {|
  date: Date,
  onClick?: (date: Date) => void,
  style: any
|};

type State = void;

class CalendarDay extends React.Component<Props, State> {
  onClick = () => {
    const { onClick, date } = this.props;
    if (onClick) {
      onClick(date);
    }
  };

  render() {
    const { date, style } = this.props;
    let dayOfMonth = getDate(date);
    return <td style={style} onClick={this.onClick}>{dayOfMonth}</td>;
  }
}

export default CalendarDay;
