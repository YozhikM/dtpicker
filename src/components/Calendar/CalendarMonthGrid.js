/* @flow */

import React from 'react';
import s from './Calendar.scss';
import Day from './Day/Day';
import { startOfMonth, endOfMonth, eachDay, getDay, format } from 'date-fns';

export type Props = {
  date: Date,
  activeDates: Date,
  onClickDay?: (date: Date) => void,
  onClickMonth?: (date: Date) => void
};

type State = {
  weeks: Array<any>
};

class CalendarMonthGrid extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      weeks: []
    };
  }

  componentWillMount() {
    const { date } = this.props;
    this.setState({ weeks: this.createMonth(date) });
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.date !== this.props.date) {
      this.setState({ weeks: this.createMonth(nextProps.date) });
    }
  }

  createMonth = (date: Date) => {
    let startDateOfMonth = startOfMonth(date);
    let endDateOfMonth = endOfMonth(date);
    let currentMonth = eachDay(startDateOfMonth, endDateOfMonth);
    let firstOfMonthDay = getDay(startDateOfMonth);
    let beginOfMonth;
    firstOfMonthDay === 0 ? (beginOfMonth = 1) : (beginOfMonth = 8 - firstOfMonthDay);
    let firstWeek = currentMonth.slice(0, beginOfMonth);
    let secondWeek = currentMonth.slice(beginOfMonth, 7 + firstWeek.length);
    let thirdWeek = currentMonth.slice(beginOfMonth + 7, beginOfMonth + 14);
    let fourthWeek = currentMonth.slice(beginOfMonth + 14, beginOfMonth + 21);
    let fiveWeek = currentMonth.slice(beginOfMonth + 21, beginOfMonth + 28);
    let tailWeek = currentMonth.slice(beginOfMonth + 28);
    const weeks = [firstWeek, secondWeek, thirdWeek, fourthWeek, fiveWeek];
    if (tailWeek.length > 0) {
      weeks.push(tailWeek);
    }
    return weeks;
  };

  isSameDate(lDate: Date, rDate: Date): boolean {
    return (
      lDate.getDate() === rDate.getDate() &&
      lDate.getMonth() === rDate.getMonth() &&
      lDate.getFullYear() === rDate.getFullYear()
    );
  }

  onClickDay = (date: Date) => {
    const { onClickDay } = this.props;
    if (onClickDay) {
      onClickDay(date);
    }
  };

  onClickMonth = () => {
    const { onClickMonth } = this.props;
    if (onClickMonth) {
      onClickMonth(this.props.date);
    }
  };

  checkInvalidDate = () => {
    const { date } = this.props;
    const validFormat = format(date, 'MMMM YYYY');
    if (validFormat === 'Invalid Date') {
      return 'Данные неверны';
    }
    else return validFormat;
  };

  render() {
    const { weeks } = this.state;
    const { activeDates, date } = this.props;

    return (
      <div className={s.root}>
        <span onClick={this.onClickMonth}>{format(date, 'MMMM YYYY')}</span>
        <table>
          <thead>
            <tr>
              <th>ПН</th>
              <th>ВТ</th>
              <th>СР</th>
              <th>ЧТ</th>
              <th>ПТ</th>
              <th>СБ</th>
              <th>ВС</th>
            </tr>
          </thead>
          <tbody>
            {weeks.map((week, i) => (
              <tr key={i}>
                {week.map((date, dayOfWeek) => {
                  return (
                    <Day
                      key={dayOfWeek}
                      date={date}
                      onClick={this.onClickDay}
                      isActive={this.isSameDate(activeDates, date)}
                    />
                  );
                })}
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
}

export default CalendarMonthGrid;
