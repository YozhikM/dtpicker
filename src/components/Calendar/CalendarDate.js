/* @flow */
/* eslint-disable jsx-a11y/no-static-element-interactions */

import React from 'react';
import { setMonth, setYear, addYears, addMonths, getYear, format } from 'date-fns';
import TableSelect from '../TableSelect/TableSelect';
import CalendarMonthGrid from './CalendarMonthGrid';
import TimePicker from './TimePicker';
import SvgIcon from '../SvgIcon';
import s from './Calendar.scss';

type Show = 'calendar' | 'mm' | 'yy' | 'yy10';

export type Props = {|
  value: Date,
  onChange?: (value: Date) => void, // works on change day, month, year, year10, time (it only work for displaying value)
  highlight?: Array<Date> | Date,
  visibleTime?: Date,
  onSetDate?: (date: Date) => void, // user click by day in calendar
  onSetTime?: (time: Date) => void, // user click on time in calendar
  leftArrow?: boolean, // show left arrow in header (true by default)
  rightArrow?: boolean, // show right arrow in header (true by default)
  time?: boolean, // display time picker (hidden by default)
  show?: Show // display needed view type (calendar by default)
|};

type State = {|
  show: 'calendar' | 'mm' | 'yy' | 'yy10',
  value: Date,
  monthsOptions: Array<{ value: number, name: string }>
|};

class CalendarDate extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      show: this.props.show || 'calendar',
      value: this.props.value,
      monthsOptions: this.getMonthOptions()
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (this.props.value !== nextProps.value) {
      this.setState({
        value: nextProps.value
      });
    }
  }

  onSetDate = (date: Date) => {
    const { onSetDate } = this.props;
    if (onSetDate) onSetDate(date);
  };

  onSetTime = (time: Date) => {
    const { onSetTime } = this.props;
    if (onSetTime) onSetTime(time);
  };

  onChange = (value: Date) => {
    this.setState({ value }, () => {
      const { onChange } = this.props;
      if (onChange) onChange(value);
    });
  };

  onChangeMonth = (n: number) => {
    const { value } = this.state;
    const newDate = setMonth(value, n);
    this.setState({ show: 'calendar' });
    this.onChange(newDate);
  };

  onChangeYear = (n: number) => {
    const { value } = this.state;
    const newDate = setYear(value, n);
    this.setState({ show: 'mm' });
    this.onChange(newDate);
  };

  onChangeDecade = (n: number) => {
    const { value } = this.state;
    const newDate = setYear(value, n);
    this.setState({ show: 'yy' });
    this.onChange(newDate);
  };

  showCalendar = () => {
    this.setState({ show: 'calendar' });
  };

  showMonthTable = () => {
    this.setState({ show: 'mm' });
  };

  showYearTable = () => {
    this.setState({ show: 'yy' });
  };

  showDecadeTable = () => {
    this.setState({ show: 'yy10' });
  };

  incrementMonth = () => {
    const value = addMonths(this.state.value, 1);
    this.onChange(value);
  };

  decrementMonth = () => {
    const value = addMonths(this.state.value, -1);
    this.onChange(value);
  };

  increment10Years = () => {
    const { value } = this.state;
    const newDate = addYears(value, 10);
    this.onChange(newDate);
  };

  decrement10Years = () => {
    const { value } = this.state;
    const newDate = addYears(value, -10);
    this.onChange(newDate);
  };

  incrementYears = () => {
    const { value } = this.state;
    const newDate = addYears(value, 1);
    this.onChange(newDate);
  };

  decrementYears = () => {
    const { value } = this.state;
    const newDate = addYears(value, -1);
    this.onChange(newDate);
  };

  showDecade = () => {
    const { value } = this.state;
    const tmp = getYear(value);
    const startYear = tmp - tmp % 10;
    const decade = `${startYear} - ${startYear + 9}`;
    return decade;
  };

  getMonthOptions() {
    return [...Array(12)].map((v, i) => ({
      value: i,
      name: format(new Date(2000, i, 1), 'MMM')
    }));
  }

  getYearOptions(date: Date) {
    const tmp = date.getFullYear();
    const startYear = tmp - tmp % 10;
    return [...Array(10)].map((v, i) => ({
      value: startYear + i,
      name: `${startYear + i}`
    }));
  }

  getYearDecadeOptions(date: Date, cnt: number = 5) {
    const tmp = date.getFullYear();
    const middleYear = tmp - tmp % 10;
    const startYear = middleYear - Math.ceil(cnt / 2) * 10;
    return [...Array(cnt)].map((v, i) => {
      const year = startYear + i * 10;
      return {
        value: year,
        name: `${year} ${year + 9}`
      };
    });
  }

  render() {
    const { show, value, monthsOptions } = this.state;
    const { time, highlight, visibleTime } = this.props;
    const leftArrowStyle = {
      visibility: this.props.leftArrow === false ? 'hidden' : 'visible'
    };
    const rightArrowStyle = {
      visibility: this.props.rightArrow === false ? 'hidden' : 'visible'
    };

    if (show === 'yy10') {
      const curYear = getYear(value);
      return (
        <div className={s.calendar_container}>
          <div className={s.table}>
            <span>
              &nbsp; <br /> &nbsp;
            </span>
            <TableSelect
              options={this.getYearDecadeOptions(value, 8)}
              cols={4}
              value={curYear - curYear % 10}
              onChange={this.onChangeDecade}
            />
            <button onClick={this.showYearTable}>Назад</button>
          </div>
        </div>
      );
    }

    if (show === 'yy') {
      return (
        <div className={s.calendar_container}>
          <div className={s.table}>
            <div className={s.c_button_container}>
              <button onClick={this.decrement10Years}>
                <SvgIcon file="arrow-left" />
              </button>
              <span onClick={this.showDecadeTable}>{this.showDecade()}</span>
              <button onClick={this.increment10Years}>
                <SvgIcon file="arrow-right" />
              </button>
            </div>
            <TableSelect
              options={this.getYearOptions(value)}
              cols={4}
              value={getYear(value)}
              onChange={this.onChangeYear}
            />
            <button onClick={this.showMonthTable}>Назад</button>
          </div>
        </div>
      );
    }

    if (show === 'mm') {
      return (
        <div className={s.calendar_container}>
          <div className={s.table}>
            <div className={s.c_button_container}>
              <button onClick={this.decrementYears}>
                <SvgIcon file="arrow-left" />
              </button>
              <span onClick={this.showYearTable}>{getYear(value)}</span>
              <button onClick={this.incrementYears}>
                <SvgIcon file="arrow-right" />
              </button>
            </div>
            <TableSelect
              options={monthsOptions}
              cols={4}
              value={value.getMonth()}
              onChange={this.onChangeMonth}
            />
            <button onClick={this.showCalendar}>Назад</button>
          </div>
        </div>
      );
    }

    return (
      <div className={s.calendar_container}>
        <button style={leftArrowStyle} className={s.left_arrow} onClick={this.decrementMonth}>
          <SvgIcon file="arrow-left" />
        </button>
        <button style={rightArrowStyle} className={s.right_arrow} onClick={this.incrementMonth}>
          <SvgIcon file="arrow-right" />
        </button>
        <CalendarMonthGrid
          highlight={highlight}
          value={value}
          onClickMonth={this.showMonthTable}
          onSetDate={this.onSetDate}
        />
        {time && <TimePicker date={visibleTime} onSetTime={this.onSetTime} />}
      </div>
    );
  }
}

export default CalendarDate;
