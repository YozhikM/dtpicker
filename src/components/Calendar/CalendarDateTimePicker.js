/* @flow */

import React from 'react';
import CalendarDate from '../Calendar/CalendarDate';
import MaskedInput from 'react-text-mask';
import SvgIcon from '../SvgIcon';
import s from './Calendar.scss';
import { format, isValid, isSameDay } from 'date-fns';

type Props = {|
  highlight: Array<Date> | Date,
  value: Date,
  visibleDate?: Date,
  time?: boolean,
  onChangeCalendarVisibility?: boolean => void,
  isCalendarShown?: boolean,
  onChange?: (value: Date) => void,
  borderLeft?: boolean,
  borderRight?: boolean,
  icon?: boolean,
  onSetDate: (date: Date) => void,
  leftArrow?: boolean,
  rightArrow?: boolean,
|};
type State = {|
  highlight: Array<Date> | Date,
  value: Date,
  visibleDate: Date,
  isCalendarShown: boolean,
  inputValue: string,
|};

class CalendarDateTimePicker extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      highlight: this.props.highlight,
      visibleDate: this.props.visibleDate,
      value: this.props.value,
      isCalendarShown: !!this.props.isCalendarShown,
      inputValue: '',
    };
  }

  formatWithTime = (date: Date) => {
    return format(date, 'DD/MM/YYYY HH:mm');
  };
  formatWithoutTime = (date: Date) => {
    return format(date, 'DD/MM/YYYY');
  };

  componentWillMount() {
    const { visibleDate, time } = this.props;
    if (time) {
      this.setState({ inputValue: this.formatWithTime(visibleDate) });
    } else {
      this.setState({
        inputValue: this.formatWithoutTime(visibleDate),
      });
    }
  }

  componentWillReceiveProps(nextProps: Props) {
    const { time } = this.props;
    if (nextProps.time !== time) {
      if (time) {
        this.setState({
          inputValue: this.formatWithTime(nextProps.visibleDate),
        });
      } else {
        this.setState({
          inputValue: this.formatWithoutTime(nextProps.visibleDate),
        });
      }
    }
    if (
      nextProps.visibleDate !== this.state.visibleDate &&
      !isSameDay(nextProps.visibleDate, this.props.value)
    ) 
		this.setState({ visibleDate: nextProps.visibleDate });
	{
      if (time) {
        this.setState({
          inputValue: this.formatWithTime(nextProps.visibleDate),
        });
      } else {
        this.setState({
          inputValue: this.formatWithoutTime(nextProps.visibleDate),
        });
      }
    }

    if (nextProps.value !== this.state.value) {
      this.setState({ value: nextProps.value });
    }

    if (nextProps.isCalendarShown !== this.state.isCalendarShown) {
      this.setState({ isCalendarShown: nextProps.isCalendarShown });
    }
  }

  onChangeInputValue = (e: any) => {
    const { onChange, onSetDate, time } = this.props;
    const value = e.target.value;
    this.setState({ inputValue: value });
    let userDate;
    const YY = Number(value.substr(6, 4));
    const MM = Number(value.substr(3, 2)) - 1;
    const DD = Number(value.substr(0, 2));
    const HH = Number(value.substr(11, 2));
    const mm = Number(value.substr(14, 2));
    if (time) {
      userDate = new Date(YY, MM, DD, HH, mm);
    } else {
      userDate = new Date(YY, MM, DD);
    }
    if (value.length > 0 && onSetDate && isValid(userDate)) {
      onSetDate(userDate);
      onChange(userDate);
    }
  };

  onFocusInput = () => {
    const { isCalendarShown } = this.state;
    if (!isCalendarShown) {
      this.onChangeVisibility();
    }
  };

  onPressEnter = (e: Event) => {
    if (e.key === 'Enter') {
      this.onChangeVisibility();
    }
  };

  onChangeVisibility = () => {
    const { isCalendarShown } = this.state;
    const { onChangeCalendarVisibility } = this.props;
    if (onChangeCalendarVisibility) {
      onChangeCalendarVisibility(!isCalendarShown);
    }
  };

  onChange = (value: Date) => {
    const { onChange } = this.props;
    if (onChange) {
      onChange(value);
    }
  };

  onSetDate = (date: Date) => {
    const { time } = this.props;
    if (time) {
      this.setState({
        inputValue: this.formatWithTime(date),
      });
    } else {
      this.setState({
        inputValue: this.formatWithoutTime(date),
      });
    }
    const { onSetDate } = this.props;
    if (onSetDate) onSetDate(date);
  };

  onSetTime = (time: Date) => {
    this.setState({
      visibleDate: time,
      inputValue: this.formatWithTime(time),
    });
  };

  render() {
    const {
      highlight,
      borderLeft,
      borderRight,
      icon,
      leftArrow,
      rightArrow,
      time,
      isCalendarShown,
      visibleDate,
	  isSingleCalendar
    } = this.props;
    const { inputValue, value } = this.state;
    const iconStyle = {
      display: icon ? 'flex' : 'none',
    };
    const borderStyle = {
      borderLeft: borderLeft ? '3px solid #34495e' : 0,
      borderRight: borderRight ? '3px solid #34495e' : 0,
    };
    let mask;
    if (time) {
      mask = [
        /[0-9]/,
        /\d/,
        '/',
        /\d/,
        /\d/,
        '/',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
        ' ',
        /\d/,
        /\d/,
        ':',
        /\d/,
        /\d/,
      ];
    } else {
      mask = [
        /[0-9]/,
        /\d/,
        '/',
        /\d/,
        /\d/,
        '/',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
      ];
    }

    return (
      <div className={s.container}>
        <MaskedInput
          mask={mask}
          className={s.input}
          style={borderStyle}
          type="text"
          onChange={this.onChangeInputValue}
          onFocus={this.onFocusInput}
          onKeyPress={this.onPressEnter}
          value={inputValue}
          showMask
        />
        <div
          onClick={this.onChangeVisibility}
          className={s.icon}
          style={iconStyle}
        >
          <SvgIcon file="calendar" />
        </div>
        {isCalendarShown && (
          <CalendarDate
            highlight={highlight}
            value={value}
            visibleTime={visibleDate}
            onSetDate={this.onSetDate}
            onSetTime={this.onSetTime}
            onChange={this.onChange}
            leftArrow={leftArrow}
            rightArrow={rightArrow}
            time={time}
			isSingleCalendar={isSingleCalendar}
          />
        )}
      </div>
    );
  }
}

export default CalendarDateTimePicker;
