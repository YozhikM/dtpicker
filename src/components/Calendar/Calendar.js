/* @flow */

import React from 'react';
import CalendarDateTimePicker from './CalendarDateTimePicker';
import SvgIcon from '../SvgIcon';
import s from './MainCalendar.scss';
import { addMonths } from 'date-fns';

type Props = {|
  highlight?: Array<Date> | Date,
  value?: Date
|};
type State = {|
  highlight: Array<Date> | Date,
  value: Date,
  time: boolean,
  isCalendarShown: boolean,
  isSingleCalendar: boolean,
  firstClick?: boolean,
  endClick?: boolean,
  firstDate?: Date,
  secondDate?: Date
|};

class Calendar extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      value: this.props.value || new Date(),
      highlight: this.props.highlight || new Date(),
      time: false,
      isCalendarShown: true,
      isSingleCalendar: true,
      firstClick: false,
      endClick: false
    };
  }

  componentWillMount() {
    const { highlight } = this.props;
    const { isSingleCalendar, value } = this.state;
    if (!isSingleCalendar && Array.isArray(highlight)) {
      this.setState({
        highlight: highlight || [value],
        firstDate: highlight[0] || value,
        secondDate: highlight[0] || value
      });
    } else {
      this.setState({ highlight: this.props.highlight || value });
    }
  }

  onSetDate = (date: Date) => {
    const { firstClick, highlight, isSingleCalendar } = this.state;

    if (!isSingleCalendar && Array.isArray(highlight)) {
      const highlightCopy = highlight.slice();
      highlightCopy.push(date);

      if (highlightCopy.length > 2) {
        highlightCopy.shift();
      }

      if (!firstClick) {
        this.setState({
          firstClick: true,
          highlight: highlightCopy,
          endClick: false,
          firstDate: highlightCopy[1]
        });
      }

      if (firstClick) {
        this.setState({
          endClick: true,
          highlight: highlightCopy,
          firstClick: false,
          secondDate: highlightCopy[1]
        });
      }
    } else {
      this.setState({ highlight: date });
    }
  };

  onChange = (value: Date) => {
    this.setState({ value });
  };

  onChangeSecondValue = (value: Date) => {
    this.setState({ value: addMonths(value, -1) });
  };

  onChangeCalendarVisibility = () => {
    const { isCalendarShown } = this.state;
    this.setState({ isCalendarShown: !isCalendarShown });
  };

  render() {
    const {
      time,
      value,
      highlight,
      isCalendarShown,
      isSingleCalendar,
      firstDate,
      secondDate
    } = this.state;
    return (
      <div>
        {isSingleCalendar ? (
          <div className={s.container}>
            <CalendarDateTimePicker
              value={value}
              highlight={highlight}
              visibleDate={highlight}
              isCalendarShown={isCalendarShown}
              onSetDate={this.onSetDate}
              onChange={this.onChange}
              onChangeCalendarVisibility={this.onChangeCalendarVisibility}
              borderLeft
              borderRight
              icon={true}
              time={time}
            />
          </div>
        ) : (
          <div className={s.container}>
            <CalendarDateTimePicker
              value={value}
              highlight={highlight}
              visibleDate={secondDate}
              isCalendarShown={isCalendarShown}
              onSetDate={this.onSetDate}
              onChange={this.onChange}
              onChangeCalendarVisibility={this.onChangeCalendarVisibility}
              borderLeft
              borderRight={false}
              rightArrow={false}
              icon={false}
              time={time}
            />
            <div className={s.arrow}>
              {secondDate && firstDate && firstDate < secondDate ? (
                <SvgIcon file="arrow-right" />
              ) : (
                <SvgIcon file="arrow-left" />
              )}
            </div>
            <CalendarDateTimePicker
              value={addMonths(value, 1)}
              highlight={highlight}
              visibleDate={firstDate}
              isCalendarShown={isCalendarShown}
              onSetDate={this.onSetDate}
              onChange={this.onChangeSecondValue}
              onChangeCalendarVisibility={this.onChangeCalendarVisibility}
              borderLeft={false}
              borderRight
              icon={false}
              leftArrow={false}
              time={time}
            />
          </div>
        )}
      </div>
    );
  }
}

export default Calendar;
