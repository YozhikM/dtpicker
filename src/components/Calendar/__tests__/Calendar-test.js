/* @flow */

import React from 'react';
import Calendar from '../Calendar';

describe('Calendar', () => {
  describe('render()', () => {
    it('basic', () => {
      const wrapper = shallow(
        <Calendar value={new Date(2018, 3)} highlight={new Date(2018, 3, 4)} />
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('!isSingleCalendar', () => {
      const wrapper = shallow(
        <Calendar
          highlight={[new Date(2009, 3, 4), new Date(2009, 3, 6)]}
          value={new Date(2009, 4)}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('componentWillMount()', () => {
    it('isSingleCalendar', () => {
      const wrapper = shallow(
        <Calendar value={new Date(2006, 7)} highlight={new Date(2006, 7, 4)} />
      );
      wrapper.setState({ isSingleCalendar: true });
      expect(wrapper).toMatchSnapshot();
    });

    it('!isSingleCalendar', () => {
      const wrapper = shallow(
        <Calendar
          value={new Date(2006, 4)}
          highlight={[new Date(2006, 3, 1), new Date(2006, 3, 1)]}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('onSetDate()', () => {
    it('isSingleCalendar', () => {
      const wrapper = shallow(
        <Calendar value={new Date(2017, 11)} highlight={new Date(2017, 11, 17)} />
      );
      wrapper.setState({ isSingleCalendar: true });
      const newDate = new Date(2009, 3, 9);
      wrapper.instance().onSetDate(newDate);
      expect(wrapper.state('highlight')).toMatchSnapshot();
    });

    describe('!isSingleCalendar', () => {
      it('!firstClick', () => {
        const wrapper = shallow(
          <Calendar
            value={new Date(2009, 4)}
            highlight={[new Date(2009, 3, 4), new Date(2009, 3, 6)]}
          />
        );
        const newDate = new Date(2009, 3, 9);
        wrapper
          .find('CalendarDateTimePicker')
          .at(0)
          .props()
          .onSetDate(newDate);
        expect(wrapper.state('highlight')).toMatchSnapshot();
      });

      it('firstClick', () => {
        const wrapper = shallow(
          <Calendar value={new Date(2009, 4)} highlight={[new Date(2009, 3, 4), new Date(2009, 3, 7)]} />
        );
        const newDate = new Date(2009, 3, 9);
        wrapper.instance().onSetDate(newDate);
        expect(wrapper.state('highlight')).toMatchSnapshot();
      });
    });
  });

  it('onChange()', () => {
    const wrapper = shallow(
      <Calendar value={new Date(2007, 6)} highlight={new Date(2009, 6, 14)} />
    );
    const newValue = new Date(2016, 4);
    wrapper.instance().onChange(newValue);
    expect(wrapper.state('value')).toMatchSnapshot();
  });

  it('onChangeSecondValue()', () => {
    const wrapper = shallow(
      <Calendar value={new Date(2007, 6)} highlight={new Date(2009, 6, 14)} />
    );
    const newValue = new Date(2016, 4);
    wrapper.instance().onChange(newValue);
    expect(wrapper.state('value')).toMatchSnapshot();
  });

  it('onChangeCalendarVisibility()', () => {
    const wrapper = shallow(
      <Calendar value={new Date(2011, 11)} highlight={new Date(2011, 11, 22)} />
    );
    wrapper.instance().onChangeCalendarVisibility();
    expect(wrapper.state('isCalendarShown')).toBe(false);
  });
});
