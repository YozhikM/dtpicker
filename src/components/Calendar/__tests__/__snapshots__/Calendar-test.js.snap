// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar componentWillMount() !isSingleCalendar 1`] = `
<div>
  <div>
    <CalendarDateTimePicker
      borderLeft={true}
      borderRight={true}
      highlight={
        Array [
          2006-03-31T18:00:00.000Z,
          2006-03-31T18:00:00.000Z,
        ]
      }
      icon={true}
      isCalendarShown={true}
      onChange={[Function]}
      onChangeCalendarVisibility={[Function]}
      onSetDate={[Function]}
      time={false}
      value={2006-04-30T18:00:00.000Z}
      visibleDate={
        Array [
          2006-03-31T18:00:00.000Z,
          2006-03-31T18:00:00.000Z,
        ]
      }
    />
  </div>
</div>
`;

exports[`Calendar componentWillMount() isSingleCalendar 1`] = `
<div>
  <div>
    <CalendarDateTimePicker
      borderLeft={true}
      borderRight={true}
      highlight={2006-08-03T18:00:00.000Z}
      icon={true}
      isCalendarShown={true}
      onChange={[Function]}
      onChangeCalendarVisibility={[Function]}
      onSetDate={[Function]}
      time={false}
      value={2006-07-31T18:00:00.000Z}
      visibleDate={2006-08-03T18:00:00.000Z}
    />
  </div>
</div>
`;

exports[`Calendar onChange() 1`] = `2016-04-30T18:00:00.000Z`;

exports[`Calendar onChangeSecondValue() 1`] = `2016-04-30T18:00:00.000Z`;

exports[`Calendar onSetDate() !isSingleCalendar !firstClick 1`] = `2009-04-08T18:00:00.000Z`;

exports[`Calendar onSetDate() !isSingleCalendar firstClick 1`] = `2009-04-08T18:00:00.000Z`;

exports[`Calendar onSetDate() isSingleCalendar 1`] = `2009-04-08T18:00:00.000Z`;

exports[`Calendar render() !isSingleCalendar 1`] = `
<div>
  <div>
    <CalendarDateTimePicker
      borderLeft={true}
      borderRight={true}
      highlight={
        Array [
          2009-04-03T18:00:00.000Z,
          2009-04-05T18:00:00.000Z,
        ]
      }
      icon={true}
      isCalendarShown={true}
      onChange={[Function]}
      onChangeCalendarVisibility={[Function]}
      onSetDate={[Function]}
      time={false}
      value={2009-04-30T18:00:00.000Z}
      visibleDate={
        Array [
          2009-04-03T18:00:00.000Z,
          2009-04-05T18:00:00.000Z,
        ]
      }
    />
  </div>
</div>
`;

exports[`Calendar render() basic 1`] = `
<div>
  <div>
    <CalendarDateTimePicker
      borderLeft={true}
      borderRight={true}
      highlight={2018-04-03T18:00:00.000Z}
      icon={true}
      isCalendarShown={true}
      onChange={[Function]}
      onChangeCalendarVisibility={[Function]}
      onSetDate={[Function]}
      time={false}
      value={2018-03-31T18:00:00.000Z}
      visibleDate={2018-04-03T18:00:00.000Z}
    />
  </div>
</div>
`;
